### DISCUSSION TITLE
Metafile (WMF / EMF) support

### DISCUSSION THREAD

**Discussion by manfromarce** [SixLabors/ImageSharp#2837] [Ideas]
Is there any plan for metafile decoding support? While .wmf and .emf are rare as standalone files, they are still common as embedded images inside Office documents and RTF. I'm currently working on a project which converts between document formats and the lack of support for handling metafiles is the only thing which prevents replacing System.Drawing.Common with a cross-platform alternative completely. Magick.NET and SkiaSharp also don't support these on non-Windows platforms but they rely on native libraries, while I'm confident ImageSharp can accomplish this.

UPDATE: [This library](https://github.com/wieslawsoltes/wmf) could be used as a starting point.

**Comment by JimBobSquarePants** [comment]
If that's something you'd be happy to make a start contributing ,I'd be happy to help out.

**Reply by manfromarce** [reply]
I would like to do this, the problem is that metafiles are made of records to be drawn, so ImageSharp.Drawing would be needed. Also WMF files aren't really device-independent and handling them is complex, I would start with EMF only.

**Reply by JimBobSquarePants** [reply]
That would be great! The decoders can be added to ImageSharp.Drawing then.  I need to do some work to update the drawing library to work with the latest ImageShap v4 builds in main here though first. 

Regarding WMF. It seems to me that if we passed a specific DPI value as decoder options that would allow us to implement the format decoder in a device independent manner?

**Reply by manfromarce** [reply]
Yes, the dimensions are expressed in twips (1/1440 inch) with a scale factor but there is no embedded DPI value. 
So should the decoder have similar APIs to other formats but in the ImageSharp.Drawing package?

**Reply by JimBobSquarePants** [reply]
Yeah it would. I have a public abstract [base class](https://github.com/SixLabors/ImageSharp/blob/4be26452b9366b0a4fe67aab91cfcb360916c5ea/src/ImageSharp/Formats/SpecializedImageDecoder%7BT%7D.cs) in main that would act as the foundation for your decoder. It's designed for passing custom decoding options. 

I'm going to start looking at ImageSharp.Drawing to see if I can get it updated to support the new code.

**Reply by lindexi** [reply]
@manfromarce Awesome, thank you. Can I ask the progress?

**Reply by manfromarce** [reply]
> @manfromarce Awesome, thank you. Can I ask the progress?

Unfortunately, I haven't had much time in recent months and have only set up the structure, but I still intend to continue.

**Reply by lindexi** [reply]
Thank you very much. Looking forward to your code.

**Reply by lindexi** [reply]
@manfromarce I write some render wmf code in https://github.com/lindexi/lindexi_gd/blob/50261e270fe9099e394139af1bb4921e937614e8/Workbench/Wmf/KawkaywecaljalKayjaybelfuber/Program.cs

But my code just the demo code. And I use skia to render not ImageSharp.

**Reply by manfromarce** [reply]
@lindexi thank you, I will look into it.

**Reply by lindexi** [reply]
@manfromarce I update it to https://github.com/lindexi/lindexi_gd/tree/master/Workbench/Wmf/SkiaWmfRenderer

**Reply by lindexi** [reply]
Hi @manfromarce I add it to my DocumentFormat.OpenXml.Extensions project, see https://github.com/dotnet-campus/DocumentFormat.OpenXml.Extensions/pull/54

**Reply by lindexi** [reply]
Hi @manfromarce I used my code to the business appliction this week. My solution is try call Inkscape first, then my code, then libwmf , see https://github.com/dotnet-campus/DocumentFormat.OpenXml.Extensions/blob/79328a88ca24bbc4d80287def31c084b3a124aa2/src/MediaConverters/SkiaWmfRenderer/src/SkiaWmfRenderer/Optimizations/EnhancedGraphicsMetafileOptimization.cs
